### PR TITLE
HLRC: Add support for XPack Post Start Basic Licence API

### PR DESCRIPTION
**Pull request by vladimirdolzhenko**
HLRC: Add support for XPack Post Start Basic Licence API

Relates to #29827

### CODE REVIEW

**Review by imotov — APPROVED** [review]
LGTM, assuming that you will fix the formatting issue and the test pass. Maybe @hub-cap should give it a look as well.

**Comment by vladimirdolzhenko** [comment]
@elasticmachine test this please

**Comment by vladimirdolzhenko** [comment]
@hub-cap could you please have a look as well ?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/RestPostStartBasicLicense.java
Comment (hub-cap) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/RestPostStartBasicLicense.java]: Pls dont change the getter here

File: x-pack/plugin/core/src/main/java/org/elasticsearch/protocol/xpack/license/PostStartBasicResponse.java
Comment (hub-cap) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/protocol/xpack/license/PostStartBasicResponse.java]: With the new way we are not sharing request/response, there is no reason to touch these, nor more them into protocol. These PRs now should be touching nothing in `x-pack` or `server`.

**Review by hub-cap — CHANGES_REQUESTED** [review]
Lets drop the word post from all this, as we want to keep it to just the name of the action. Of course, in the request converters we will know its a POST, but the HL client should be agnostic of this. Also please move the stuff you put back in `protocol` back. Sorry about any confusion, the rest client has been in lots of flux.

**Comment by hub-cap** [comment]
as mentioned in https://github.com/elastic/elasticsearch/pull/33406#pullrequestreview-156178012, we need to add a `LicensingIT` test class for testing randomized values in these calls, and negative tests. Please add that. I wanted to just drop a comment before I started reviewing so you know of this quicker.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/license/PostStartBasicRequest.java
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/license/PostStartBasicRequest.java]: pls make this boolean final and remove the setter.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/license/PostStartBasicResponse.java
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/license/PostStartBasicResponse.java]: This class looks like it has a lot of extra stuff that is not needed. In general, we do not need from/to transport serialization. Do we need addCustomFields, as its also toXContent related.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/license/PostStartBasicResponse.java
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/license/PostStartBasicResponse.java]: see https://github.com/elastic/elasticsearch/pull/33406/files#r218858319

File: client/rest-high-level/src/main/java/org/elasticsearch/client/license/PostStartBasicResponse.java
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/license/PostStartBasicResponse.java]: is this a more common class that can be used elsewhere? if so, lets break it out to its own class, and put it in the same package.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/license/PostStartBasicResponseTests.java
Comment (hub-cap) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/license/PostStartBasicResponseTests.java]: These will have to change as well, once you remove the ToXContent portion of the response. You will have to manually construct the parser and then pass that in to the fromXContent of the response.

File: docs/java-rest/high-level/licensing/post-start-basic.asciidoc
Comment (hub-cap) [inline docs/java-rest/high-level/licensing/post-start-basic.asciidoc]: You may want to lift some of the wording from the *very* similar PR ive linked here, https://github.com/elastic/elasticsearch/blob/07554dda3b4bb2e2871d913f00e3cead0cedeeb4/docs/java-rest/high-level/licensing/start-trial.asciidoc .

**Review by hub-cap — CHANGES_REQUESTED** [review]
Great work so far. Some stuff to address below, nothing huge.

**Comment by hub-cap** [comment]
Pls also add a LicenseRequestConvertersTests to test your request converter.

**Comment by vladimirdolzhenko** [comment]
@elasticmachine test this please

**Comment by vladimirdolzhenko** [comment]
@elasticmachine test this please

**Comment by vladimirdolzhenko** [comment]
@elasticmachine test this please

File: client/rest-high-level/src/main/java/org/elasticsearch/client/license/StartBasicRequest.java
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/license/StartBasicRequest.java]: You can leave this out if you are not using it

**Review by hub-cap — APPROVED** [review]
LGTM, one nit then you can merge

File: client/rest-high-level/src/main/java/org/elasticsearch/client/license/StartBasicRequest.java
Comment (vladimirdolzhenko) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/license/StartBasicRequest.java]: with ccbc485 `StartBasicRequest` extends `TimedRequest`, added couple test routines for that

**Comment by vladimirdolzhenko** [comment]
there is a dependency for tests on https://github.com/elastic/elasticsearch/pull/33406 - I'd rather wait for start trial license will be merged

**Comment by hub-cap** [comment]
if you want to merge, you can always just stub in some Low level rest calls and then @andyb-elastic can clean them up in his PR ;)

**Comment by vladimirdolzhenko** [comment]
@hub-cap thanks for the tip: in fact `start trial license` did not help as it had been already started. 

there are several ways (at least 2) on how to solve the problem: 
- put a hard coded trial license (as it is not possible to bring license-tools to hlrc due to jar hell) - it could works for snapshot build only
- split license specific tests into a separated gradle task
- any other 

for now I rely on a hard coded trial license approach

**Comment by hub-cap** [comment]
Ahh right it was *this* issue, and I discussed with @andyb-elastic and he is going to make a new PR with the changes to make a new cluster with the different license test requirements. So it might make sense to merge this and backport it and await his change, so you can also make your change. Your call.

**Comment by andyb-elastic** [comment]
Right, looks like you got the case where it starts a basic license working by manually setting a trial license. I'm working on a PR to run the license IT in a separate gradle project (we ran into some issues getting it to work with multiple clusters in the same project)

**Pull request merged**
